streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -505,11 +509,17 @@ public void shutdown(final boolean clean) {
 
     @Override
     public boolean isEmpty() {
-        return super.isEmpty()
-            && restoring.isEmpty()
-            && restoringByPartition.isEmpty()
-            && restoredPartitions.isEmpty()
-            && suspended.isEmpty();
+        if (restoring.isEmpty() && !restoringByPartition.isEmpty()) {
+            log.error("Assigned stream tasks in an inconsistent state: the set of restoring tasks is empty but the " +
+                      "restoring by partitions map contained {}", restoringByPartition);
+            return false;
```

Comment:
It is save to return `false`? Should we rather throw an `IllegalStateException` ?

Reply:
+1, we should just throw in this case.

Reply:
Well, yeah we should

Reply:
Oh right I remember why I didn't want to just throw outright here, because we want to check all three objects for emptiness in `TaskManager#closeLostTasks` and then log which weren't empty, and what was in them (as you commented below). And it would be awkward if the `active.isEmpty` condition itself threw, since we're testing that condition to determine if we need to throw a IllegalStateException in the first place.

Does that make sense?

Reply:
I do see the motivation for throwing, since the `isEmpty` methods may be re-used in the future for cases other than throwing an `IllegalStateException` anyways. 
On the other hand, it really is awkward to throw an exception from an `isEmpty` conditional and I'm not sure a future user of this method would be expecting it to throw an exception, or notice that it does... I'm still open to throwing, just laying out my concern

Reply:
I did change it to throw though, so if you're satisfied then we can merge -- once the final system test run passes so we don't go through this again :) 

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3363/